axiom/auth/auth.go

```diff
@@ -0,0 +1,180 @@
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		return "", err
+	}
+	defer lis.Close()
+
+	redirectURL, err := url.Parse("http://" + lis.Addr().String())
+	if err != nil {
+		return "", err
+	}
+
+	config := &oauth2.Config{
+		ClientID: clientID,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:   authURL.String(),
+			TokenURL:  tokenURL.String(),
+			AuthStyle: oauth2.AuthStyleInParams,
+		},
+		RedirectURL: redirectURL.String(),
+		Scopes:      []string{"*"},
+	}
+
+	// Create the PKCE Code Verifier and S256 Code Challenge.
+	method := pkce.MethodS256
+	codeVerifier, err := pkce.New()
+	if err != nil {
+		return "", err
+	}
+
+	state, err := uuid.NewUUID()
+	if err != nil {
+		return "", err
+	}
+
+	loginURL := config.AuthCodeURL(state.String(),
+		codeVerifier.Challenge(method).AuthCodeOption(),
+		method.AuthCodeOption(),
+	)
+
+	if err = loginFunc(ctx, loginURL); err != nil {
+		return "", err
+	}
+
+	var (
+		code   string
+		doneCh = make(chan struct{})
+	)
+	srv := http.Server{
+		Addr:        lis.Addr().String(),
+		Handler:     callbackHandler(&code, state.String(), doneCh),
+		BaseContext: func(net.Listener) context.Context { return ctx },
+	}
+	defer srv.Close()
+
+	errCh := make(chan error)
+	go func() {
+		if serveErr := srv.Serve(lis); serveErr != nil && serveErr != http.ErrServerClosed {
+			errCh <- serveErr
+		}
+		close(errCh)
+	}()
+
+	select {
+	case <-ctx.Done():
+		close(doneCh)
+		return "", ctx.Err()
+	case err = <-errCh:
+		close(doneCh)
+		return "", err
+	case <-doneCh:
+		if shutdownErr := srv.Shutdown(ctx); shutdownErr != nil {
+			return "", err
+		}
+	}
+
+	token, err := config.Exchange(ctx, code, codeVerifier.AuthCodeOption())
+	if err != nil {
+		return "", err
+	}
+
+	return token.AccessToken, nil
+}
+
+func callbackHandler(code *string, state string, doneCh chan<- struct{}) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodGet {
+			code := http.StatusMethodNotAllowed
+			http.Error(w, http.StatusText(code), code)
+			return
+		}
+
+		// Make sure state matches.
+		if r.FormValue("state") != state {
+			code := http.StatusBadRequest
+			http.Error(w, http.StatusText(code), code)
+			return
+		}
+
+		tmpl, err := template.New("callback").Parse(callbackTmpl)
+		if err != nil {
+			code := http.StatusInternalServerError
+			http.Error(w, http.StatusText(code), code)
+			return
+		}
+
+		data := make(map[string]string)
+		if oauthErr := r.FormValue("error"); oauthErr != "" {
+			data["error"] = oauthErr
+			data["error_description"] = r.FormValue("error_description")
+		} else if *code = r.FormValue("code"); *code == "" {
+			code := http.StatusBadRequest
+			http.Error(w, http.StatusText(code), code)
+			return
+		}
+
+		var buf bytes.Buffer
+		if err = tmpl.Execute(&buf, data); err != nil {
+			code := http.StatusInternalServerError
+			http.Error(w, http.StatusText(code), code)
+			return
+		}
+
+		_, _ = io.Copy(w, &buf)
```

Comment:
This can err

Reply:
I'm open for suggestions. I didn't want error handling to become to crazy in the callback and possible deal with another channel, here.

Reply:
Deprecated.